pages/quests/index.tsx

```diff
@@ -8,24 +8,139 @@ import {
   const { hasNFT } = useHasNFT()
+  const { activeChain } = useNetwork()
+  const kamonNFT = getContractAddress({
+    name: 'kamonNFT',
+    chainId: activeChain?.id
+  })
+  const { balanceOf } = useBalanceOf(kamonNFT, data?.address)
+  const { tokenIdOf } = useTokenIdOf(kamonNFT, data?.address)
+  const { tokenURI } = useTokenURI(kamonNFT, tokenIdOf?.toNumber() || 0)
+  const [tokenId, setTokenId] = useState<BigInt>(BigInt(0))
+  const [tokenJSON, setTokenJSON] = useState<KamonToken>()
+  const [finalTokenUri, setFinalTokenUri] = useState('')
+  const [newTokenImageURI, setNewTokenImageURI] = useState('')
+  const [newTokenJSON, setNewTokenJSON] = useState<KamonToken>()
+  const [updateTxFailed, setUpdateTxFailed] = useState<boolean>()
+
+  const { ipfsSubmit, ipfsSubmitSucceeded, ipfsSubmitIsSubmitting } = useIpfsSubmit()
+  const { ipfsGet, ipfsGetIsSubmitting } = useIpfsGet()
+  const { update, updated, updateOwnNftIsSubmitting } = useUpdateOwnNFT(
+    kamonNFT,
+    tokenId,
+    finalTokenUri,
+  )
+
+  const submittingSomething = () => {
+    return isSubmitting || ipfsSubmitIsSubmitting || ipfsGetIsSubmitting || updateOwnNftIsSubmitting
+  }
+  useEffect(() => {
+    if (balanceOf && tokenIdOf && tokenURI) {
+      const fetchData = async () => {
+        const pinataRequest = await fetch(tokenURI.toString())
+        const responseJson = await pinataRequest.json()
+        setTokenJSON(responseJson)
+      }
+      fetchData()
+    }
+  }, [balanceOf, tokenIdOf, tokenURI])
+
+  useEffect(() => {
+    if (keywordSubmitSucceeded && !ipfsSubmitIsSubmitting
+      && ipfsSubmitSucceeded == undefined
+      && tokenJSON !== undefined && data?.address !== undefined
+    ) {
+      const userAddress = data?.address
+      const ipfsSubmitWrapper = async () => {
+        const ipfsSubmitRet = await ipfsSubmit(tokenJSON, userAddress)
+        if (ipfsSubmitRet !== 'error' && ipfsSubmitRet !== 'same_points') {
+          const tempFinalTokenUri = await ipfsSubmitRet
+          setFinalTokenUri(tempFinalTokenUri)
+        }
+      }
+      setUpdateTxFailed(false)
+      ipfsSubmitWrapper()
+    }
+  }, [keywordSubmitSucceeded, data?.address, tokenJSON, ipfsSubmit, ipfsSubmitIsSubmitting, ipfsSubmitSucceeded])
+
+  useEffect(() => {
+    if (finalTokenUri && !ipfsGetIsSubmitting && newTokenJSON == undefined) {
+      const getNewToken = async () => {
+        const ipfsGetRet = await ipfsGet(finalTokenUri)
+        if (ipfsGetRet == 'error') {
+          return
+        }
+        setNewTokenJSON(ipfsGetRet)
+        const theTokenId = tokenIdOf? tokenIdOf: BigInt(0)
+        if(theTokenId == BigInt(0)) { return }
+        setTokenId(BigInt(parseInt(theTokenId.toString())))
+        setNewTokenImageURI(ipfsGetRet.image)
+      }
+      getNewToken()
+    }
+  }, [newTokenJSON, finalTokenUri, ipfsGet, ipfsGetIsSubmitting, tokenIdOf])
+
+  useEffect(() => {
```

Comment:
@SamuraiT How does this look? If we need to move more things out of this file, the next thing I would take out is this hook, with the toasts and all that.